### ISSUE TITLE
blob/azureblob: Using a SAS token results in "Signature fields not well formed"

### ISSUE THREAD

**Issue by stanhu** [google/go-cloud#3332]
### Describe the bug

I attempted to upgrade from `gocloud.dev` from v0.27.0 to v0.28.0. This also pulled in an upgrade from `github.com/Azure/azure-sdk-for-go/sdk/storage/azblob` v0.4.1 to v0.6.1.

When I attempt uploads with an Azure SAS token, I see:

```
FATAL: blob (key "runner/eehdx1xw/project/2/default-non_protected") (code=PermissionDenied): PUT https://redacted.blob.core.windows.net/runner/eehdx1xw/project/2/default-non_protected
--------------------------------------------------------------------------------
RESPONSE 403: 403 Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.
ERROR CODE: AuthenticationFailed
--------------------------------------------------------------------------------
﻿<?xml version="1.0" encoding="utf-8"?>
<Error><Code>AuthenticationFailed</Code><Message>Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.
RequestId:9bba2735-701e-0031-33ab-06298c000000
Time:2023-10-24T18:56:11.6601824Z</Message><AuthenticationErrorDetail>Signature fields not well formed.</AuthenticationErrorDetail></Error>
```

### To Reproduce

Attempt to download or upload a blob with `AZURE_STORAGE_SAS_TOKEN` and `AZURE_STORAGE_ACCOUNT`.

## Expected behavior

The upload succeeds.

### Version

This appears on `gocloud.dev` v0.28.0 to v0.34.0.

### Additional context

I believe the issue here is that GoCloud appends the SAS token query string before the container name, but it should be appended **after**. For example if `container1` is the container name, the old URL was:

`https://example.blob.core.windows.net?se=2023-10-24T19%3A13%3A55Zsig=SIGNATUREsp=w&spr=https&sr=b&st=2023-10-24T17%3A13%3A55Z&sv=2020-02-10/container1`

I verified the upload worked when I hacked my GoCloud to append the SAS query string **after** the container name:

`https://example.blob.core.windows.net/container1?se=2023-10-24T19%3A13%3A55Zsig=SIGNATUREsp=w&spr=https&sr=b&st=2023-10-24T17%3A13%3A55Z&sv=2020-02-10`

I think https://github.com/google/go-cloud/pull/3187 might have broken this with this line:

https://github.com/google/go-cloud/blob/375e097193225b4ef300a336e04ccad9ace50bc1/blob/azureblob/azureblob.go#L347

We might want to fix this by:

1. Make `ServiceURL` a `URL` type instead of a `string`
2. In `NewServiceURL()`, fill in the various parts (scheme, query parameters, path, etc.).
3. In `NewClient`, prepend `containerName` to the path and then generate a string representation.

**Comment by vangent** [comment]
Are you sure this wasn't fixed by https://github.com/google/go-cloud/pull/3284 ?

**Comment by stanhu** [comment]
Oh, perhaps it was. I thought I tested v0.34.0, but let me try again.

**Comment by stanhu** [comment]
Yes, looks like it works after all. I didn't recompile something. Thanks for the fast response.
